contrib/pyln-proto/pyln/proto/message/fundamental_types.py

```diff
@@ -0,0 +1,231 @@
+#! /usr/bin/python3
+import struct
+
+
+def split_field(s):
+    """Helper to split string into first part and remainder"""
+    def len_without(s, delim):
+        pos = s.find(delim)
+        if pos == -1:
+            return len(s)
+        return pos
+
+    firstlen = min([len_without(s, d) for d in (',', '}', ']')])
+    return s[:firstlen], s[firstlen:]
+
+
+class FieldType(object):
+    """A (abstract) class representing the underlying type of a field.
+These are further specialized.
+
+    """
+    def __init__(self, name):
+        self.name = name
+
+    def only_at_tlv_end(self):
+        """Some types only make sense inside a tlv, at the end"""
+        return False
+
+    def name_and_val(self, name, v):
+        """This is overridden by LengthFieldType to return nothing"""
+        return " {}={}".format(name, self.val_to_str(v, None))
+
+    def is_optional(self):
+        """Overridden for tlv fields and optional fields"""
+        return False
+
+    def len_fields_bad(self, fieldname, fieldvals):
+        """Overridden by length fields for arrays"""
+        return []
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return self.name
+
+
+class IntegerType(FieldType):
+    def __init__(self, name, bytelen, structfmt):
+        super().__init__(name)
+        self.bytelen = bytelen
+        self.structfmt = structfmt
+
+    def val_to_str(self, v, otherfields):
+        return "{}".format(int(v))
+
+    def val_from_str(self, s):
+        a, b = split_field(s)
+        return int(a), b
+
+    def val_to_bin(self, v, otherfields):
+        return struct.pack(self.structfmt, v)
+
+    def val_from_bin(self, bytestream, otherfields):
+        "Returns value, bytesused"
+        if self.bytelen > len(bytestream):
+            raise ValueError('{}: not enough remaining to read'.format(self))
+        return struct.unpack_from(self.structfmt,
+                                  bytestream)[0], self.bytelen
```

Comment:
I think taking a stream into these functions (which are mainly internal) is The Right Thing, so you can do it on a file etc, rather than requiring it to be in-memory?

Reply:
Correct, if we accept a file-like interface to read from it'd allow us to pass in a file / socket / in-memory buffer and just read/write from/to it, and avoid having to load in memory first.

I usually take both a string and a bytesio (breaking all kinds of typing rules, and differentiating with `instanceof`) and wrap a string in a `io.BytesIO` instance if it is a string when the interface is public facing, since not everybody knows how to wrap an in-memory buffer with it, but if this is internal only it'd make sense to use `io.BytesIO` everywhere.